toolkit/offchain/src/collateral_selection.rs

```diff
@@ -0,0 +1,74 @@
+use ogmios_client::types::*;
+
+pub const MIN_REQUIRED_COLLATERAL: u64 = 5_000_000;
+
+// Largest-First selection algorithm based on https://cips.cardano.org/cip/CIP-0002#largest-first
+pub fn largest_first(
+	inputs_available: &[OgmiosUtxo],
+	amount_required: u64,
+	max_collateral_inputs: u32,
+	min_required_collateral: u64,
+) -> Result<Vec<&OgmiosUtxo>, &str> {
+	let amount_required = amount_required.max(min_required_collateral);
+	let amount_available: u64 = inputs_available.iter().map(|utxo| utxo.value.lovelace).sum();
+	if amount_available < amount_required {
+		return Err("InputValueInsufficientError");
```

Comment:
I really suggest that error should contain user friendly error message. 
It seems there is no way to handle this situation. Error will be passed to the user. I don't see any added value in using more sophisticated type.